pkg/components/pubsub/pluggable_test.go

```diff
@@ -0,0 +1,293 @@
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+	"sync"
+	"testing"
+
+	guuid "github.com/google/uuid"
+
+	contribMetadata "github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/components-contrib/pubsub"
+
+	"go.uber.org/atomic"
+
+	"github.com/dapr/dapr/pkg/components"
+	"github.com/dapr/dapr/pkg/components/pluggable"
+	proto "github.com/dapr/dapr/pkg/proto/components/v1"
+	testingGrpc "github.com/dapr/dapr/pkg/testing/grpc"
+
+	"github.com/dapr/kit/logger"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"google.golang.org/grpc"
+)
+
+var testLogger = logger.NewLogger("pubsub-pluggable-test")
+
+type server struct {
+	proto.UnimplementedPubSubServer
+	initCalled      atomic.Int64
+	onInitCalled    func(*proto.PubSubInitRequest)
+	initErr         error
+	featuresCalled  atomic.Int64
+	featuresErr     error
+	publishCalled   atomic.Int64
+	onPublishCalled func(*proto.PublishRequest)
+	publishErr      error
+	pullChan        chan *proto.PullMessageResponse
+	pingCalled      atomic.Int64
+	pingErr         error
+	onAckReceived   func(*proto.PullMessagesRequest)
+	pullCalled      atomic.Int64
+	pullErr         error
+}
+
+//nolint:nosnakecase
+func (s *server) PullMessages(svc proto.PubSub_PullMessagesServer) error {
+	s.pullCalled.Add(1)
+
+	if s.onAckReceived != nil {
+		go func() {
+			for {
+				msg, err := svc.Recv()
+				if err != nil {
+					return
+				}
+				s.onAckReceived(msg)
+			}
+		}()
+	}
+	if s.pullChan != nil {
+		for msg := range s.pullChan {
+			if err := svc.Send(msg); err != nil {
+				return err
+			}
+		}
+	}
+
+	return s.pullErr
+}
+
+func (s *server) Init(_ context.Context, req *proto.PubSubInitRequest) (*proto.PubSubInitResponse, error) {
+	s.initCalled.Add(1)
+	if s.onInitCalled != nil {
+		s.onInitCalled(req)
+	}
+	return &proto.PubSubInitResponse{}, s.initErr
+}
+
+func (s *server) Features(context.Context, *proto.FeaturesRequest) (*proto.FeaturesResponse, error) {
+	s.featuresCalled.Add(1)
+	return &proto.FeaturesResponse{}, s.featuresErr
+}
+
+func (s *server) Publish(_ context.Context, req *proto.PublishRequest) (*proto.PublishResponse, error) {
+	s.publishCalled.Add(1)
+	if s.onPublishCalled != nil {
+		s.onPublishCalled(req)
+	}
+	return &proto.PublishResponse{}, s.publishErr
+}
+
+func (s *server) Ping(context.Context, *proto.PingRequest) (*proto.PingResponse, error) {
+	s.pingCalled.Add(1)
+	return &proto.PingResponse{}, s.pingErr
+}
+
+func TestPubSubPluggableCalls(t *testing.T) {
+	getPubSub := testingGrpc.TestServerFor(testLogger, func(s *grpc.Server, svc *server) {
+		proto.RegisterPubSubServer(s, svc)
+	}, func(cci grpc.ClientConnInterface) *grpcPubSub {
+		client := proto.NewPubSubClient(cci)
+		pubsub := fromConnector(testLogger, pluggable.NewGRPCConnector(components.Pluggable{}, proto.NewPubSubClient))
+		pubsub.Client = client
+		return pubsub
+	})
+
+	if runtime.GOOS != "windows" {
+		t.Run("test init should populate features and call grpc init", func(t *testing.T) {
```

Comment:
We should have the `else` branch for ` if runtime.GOOS != "windows"` defined here. That's something I was thinking about pluggable components and windows, beyond this test: we are not giving Windows user enough feedback that pluggable components won't run in windows. Enabling it should fail. Running this test should fail. This test should not even be ran at all in Windows IMHO.

Reply:
I think this was already aligned before I started on this feature. @johnewart had already told me that it was expected to only work on UNIX and this will be made clear at some point in the release notes. The test itself I don't see as a problem, because it is more related to us (as developers) than to the end user, and then I think there are different actions to be taken.

For the specific test I can add an "else" + a warn saying that the test will be skipped, what do you think?

The same should be done but when trying to register pluggable components though - which I can do in the upcoming PRs, is that ok for you @tmacam ?